iothub.class.nut

```diff
@@ -5,724 +5,1024 @@
-class iothub.SharedAccessSignature {
+            // If the keyName argument to the create method was falsy, skn shall not be defined.
+            // <urlEncodedKeyName> shall be the URL-encoded value of keyName.
+            // The skn property shall be the value <urlEncodedKeyName>.
+            if (keyName) sas.skn = iothub.Authorization.encodeUriComponentStrict(keyName);
 
-    sr = null;
-    sig = null;
-    skn = null;
-    se = null;
+            // The se property shall have the value of expiry.
+            sas.se = expiry;
 
-    static function create(resourceUri, keyName, key, expiry) {
+            return sas;
+        }
+
+
+        function toString() {
+            // The toString method shall return a shared-access signature token of the form:
+            // SharedAccessSignature sr=<resourceUri>&sig=<urlEncodedSignature>&se=<expiry>&skn=<urlEncodedKeyName>
+            local sas = "SharedAccessSignature ";
+            foreach (key in ["sr", "sig", "skn", "se"]) {
+                // The skn segment is not part of the returned string if the skn property is not defined.
+                if (this[key]) {
+                    if (sas[sas.len() - 1] != ' ') sas += "&";
+                    sas += key + "=" + this[key];
+                }
+            }
 
-        // The create method shall create a new instance of SharedAccessSignature with properties: sr, sig, se, and optionally skn.
-        local sas = iothub.SharedAccessSignature();
+            return sas;
+        }
+    }
 
-        // The sr property shall have the value of resourceUri.
-        sas.sr = resourceUri;
+    Endpoint = class {
 
-        // <signature> shall be an HMAC-SHA256 hash of the value <stringToSign>, which is then base64-encoded.
-        // <stringToSign> shall be a concatenation of resourceUri + "\n" + expiry.
-        local hash = iothub.Authorization.hmacHash(key, iothub.Authorization.stringToSign(resourceUri, expiry));
+        static function devicePath(id) {
+            return "/devices/" + id;
+        }
 
-        // The sig property shall be the result of URL-encoding the value <signature>.
-        sas.sig = iothub.Authorization.encodeUriComponentStrict(hash);
+        static function eventPath(id) {
+            return devicePath(id) + "/messages/events";
+        }
 
-        // If the keyName argument to the create method was falsy, skn shall not be defined.
-        // <urlEncodedKeyName> shall be the URL-encoded value of keyName.
-        // The skn property shall be the value <urlEncodedKeyName>.
-        if (keyName) sas.skn = iothub.Authorization.encodeUriComponentStrict(keyName);
+        static function messagePath(id) {
+            return devicePath(id) + "/messages/devicebound";
+        }
 
-        // The se property shall have the value of expiry.
-        sas.se = expiry;
+        static function claimBasedSecurityPath() {
+            return "$cbs";
```

Comment:
It should be constant, as it's more efficient than calling a function (less space and faster).